src/pkg/quota/service.go

```diff
@@ -0,0 +1,121 @@
+	Cpus       float32
+	Gpus       uint32
+	MemoryMiB  float32
+	Replicas   uint32
+	ShmSizeMiB float32
+}
+
+// Based on https://www.ietf.org/rfc/rfc3986.txt, using the pattern for query
+// (which is a superset of path's `pchar`) but removing the single quote.
+//
+//	query       = *( pchar / "/" / "?" )
+//	pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+//	unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+//	pct-encoded   = "%" HEXDIG HEXDIG
+//	sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
+var healthcheckUrlRegex = regexp.MustCompile(`(?i)(http://)?(localhost|127.0.0.1)(:\d{1,5})?(/(?:[?a-z0-9._~!$&()*+,;=:@-]|%[a-f0-9]{2})*)*`)
+
+func (q ServiceQuotas) Validate(service *defangv1.Service) error {
+	if service.Name == "" {
+		return errors.New("service name is required") // CodeInvalidArgument
+	}
+
+	if service.Build != nil {
+		if service.Build.Context == "" {
+			return errors.New("build.context is required") // CodeInvalidArgument
+		}
+		if service.Build.ShmSize > q.ShmSizeMiB || service.Build.ShmSize < 0 {
+			return fmt.Errorf("build.shm_size exceeds quota (max %v MiB)", q.ShmSizeMiB) // CodeInvalidArgument
+		}
+	} else {
+		if service.Image == "" {
+			return errors.New("missing image or build") // CodeInvalidArgument
+		}
+	}
+
+	// hasHost := false
+	hasIngress := false
+	uniquePorts := make(map[uint32]bool)
+	for _, port := range service.Ports {
+		if port.Target < 1 || port.Target > 32767 {
+			return fmt.Errorf("port %d is out of range", port.Target) // CodeInvalidArgument
+		}
+		if port.Mode == defangv1.Mode_INGRESS {
+			if port.Protocol == defangv1.Protocol_TCP || port.Protocol == defangv1.Protocol_UDP {
+				return fmt.Errorf("mode:INGRESS is not supported by protocol:%s", port.Protocol) // CodeInvalidArgument
+			}
+		}
+		if uniquePorts[port.Target] {
+			return fmt.Errorf("duplicate port %d", port.Target) // CodeInvalidArgument
+		}
+		// hasHost = hasHost || port.Mode == v1.Mode_HOST
+		hasIngress = hasIngress || port.Mode == defangv1.Mode_INGRESS
+		uniquePorts[port.Target] = true
+	}
+	if service.Healthcheck != nil && len(service.Healthcheck.Test) > 0 {
+		// Technically this should test for <= but both interval and timeout have 30s as the default value in compose spec
+		if service.Healthcheck.Interval > 0 && service.Healthcheck.Interval < service.Healthcheck.Timeout {
+			return errors.New("invalid healthcheck: timeout must be less than the interval")
+		}
+		switch service.Healthcheck.Test[0] {
+		case "CMD", "CMD-SHELL":
+			if hasIngress {
+				// For ingress ports, we derive the target group healthcheck path/port from the service healthcheck
+				hasHttpUrl := false
+				for _, arg := range service.Healthcheck.Test[1:] {
+					if healthcheckUrlRegex.MatchString(arg) {
+						hasHttpUrl = true
+						break
+					}
+				}
+				if !hasHttpUrl {
+					return errors.New("invalid healthcheck: ingress ports require a healthcheck with HTTP URL")
```

Comment:
I wonder if we should be more specific in this error message that we are looking for localhost URLs. I wonder if someone might try to use their service name in here or something by accident. 🤷 

```
invalid healthcheck: ingress ports require an HTTP healthcheck on `localhost` or `127.0.0.1`.
```

Would it be worth adding an example to this error, or would it be too verbose?

```
invalid healthcheck: ingress ports require an HTTP healthcheck on `localhost` or `127.0.0.1`.  For example: `["CMD", "curl", "-f", "http://localhost:5000/"]`.
```

Reply:
Good call. We should encourage `localhost` since it'll be IPv6 compatible (127.0.0.1 is not)